### ISSUE TITLE
Is there any way to receive notifications for retry attempts using the RetryPolicy?

### ISSUE THREAD

**Issue by pcwiese** [grpc/grpc-dotnet#1904]
Ideally, I would like to be able to plug into the retry loops in some way so I can emit metrics or perhaps inject a custom header indicating which retry attempt the call is for. Is there any way to do that? It all looks internal to me and the only thing I see are ILogger-based logs coming out of this stuff.

Am I missing something? Or just not possible?

**Comment by JamesNK** [comment]
A header is added to gRPC requests when a retry happens - https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exposed-retry-metadata

There is also logging when a retry happens, but it is difficult to hook into.

Is that what you're looking for?

**Comment by pcwiese** [comment]
I think it is. Sounds like I can read the injected header on the server AND in our instrumented client interceptor that sits above the call. I'll try it out

**Comment by JamesNK** [comment]
Added some docs https://github.com/dotnet/AspNetCore.Docs/pull/27223

**Comment by pcwiese** [comment]
I'd like to re-open this issue.

These headers only seem to be available if a connection is actually made. If I shut down my test server, I see the retry logic attempting socket connections according to the retry policy but the retry headers aren't available at this stage it seems. Ideally, I want to be able to plug into the call chain for either scenario.

Originally I started with our own interceptor that attempted to track rpc start/stop for each retry. That works fine for our own domain specific retry interceptors. However when using the built-in retry interceptor it would only show a single start/stop, which makes sense. The retry logic sits below any interceptors configured for the call.

Then I tried wrapping the channel's sockethttpmessagehandler with a delegating one which emitted our custom instrumentation. This approach gave me the correct number of calls for our own retry interceptors and the built-in one, however that solution fails to catch the connection level retries. 

I see GrpcChannel.cs configuring the BalancerHttpHandler as the head of the chain. My custom handler was attached to this one. There doesn't appear to be a way to get notified of the connection attempt with configuration. I think it would work if we were allowed to plug something into the head of this chain.
